python/lsst/daf/butler/registry/_registry.py

```diff
@@ -657,84 +657,86 @@ def find(self, collection: str, datasetType: Union[DatasetType, str], dataId: Op
-        # Make an expanded, standardized data ID up front, so we don't do that
-        # multiple times in calls below.  Note that calling expandDataId with a
-        # full ExpandedDataCoordinate is basically a no-op.
-        dataId = self.expandDataId(dataId, graph=datasetType.dimensions, **kwds)
-        # Retrieve the ID for the given run.
-        runId = self._getRunIdFromName(run)
-        # Add the dataset table entry itself.  Note that this will get rolled
-        # back if the subsequent call to associate raises, which is what we
-        # want.
-        datasetRef = DatasetRef(datasetType=datasetType, dataId=dataId)
-        # TODO add producer
-        row = {k.name: v for k, v in dataId.full.items()}
-        row.update(
-            dataset_type_name=datasetType.name,
-            run_id=runId,
-            dataset_ref_hash=datasetRef.hash,
-            quantum_id=None
-        )
-        datasetId, = self._db.insert(self._tables.dataset, row, returnIds=True)
-        datasetRef = datasetRef.resolved(id=datasetId, run=run)
-        # A dataset is always initially associated with its run as a
-        # collection.
-        self.associate(run, [datasetRef, ])
-        if recursive:
-            for component in datasetType.storageClass.components:
-                compTypeName = datasetType.componentTypeName(component)
-                compDatasetType = self.getDatasetType(compTypeName)
-                compRef = self.addDataset(compDatasetType, dataId, run=run, producer=producer,
-                                          recursive=True)
-                self.attachComponent(component, datasetRef, compRef)
-        return datasetRef
+        rows = []
+        refs = []
+        base = {
+            "dataset_type_name": datasetType.name,
+            "run_id": self._getRunIdFromName(run),
+            "quantum_id": producer.id if producer is not None else None,
+        }
+        # Expand data IDs and build both a list of unresolved DatasetRefs
+        # and a list of dictionary rows for the dataset table.
+        for dataId in dataIds:
+            ref = DatasetRef(datasetType, self.expandDataId(dataId, graph=datasetType.dimensions))
+            refs.append(ref)
+            row = dict(base, dataset_ref_hash=ref.hash)
+            for dimension, value in ref.dataId.full.items():
+                row[dimension.name] = value
+            rows.append(row)
+        # Actually insert into the dataset table.
+        datasetIds = self._db.insert(self._tables.dataset, *rows, returnIds=True)
+        # Resolve the DatasetRefs with the autoincrement IDs we generated.
+        refs = [ref.resolved(id=datasetId, run=run) for datasetId, ref in zip(datasetIds, refs)]
+        # Associate the datasets with the run as a collection.  Note that we
+        # do this before inserting component datasets so recursing doesn't try
+        # to associate those twice.
+        self.associate(run, refs)
```

Comment:
Im a bit worried this is a non-atomic in so far as the data get inserted but things can get interrupted such that they never end up being associated

Reply:
The `transactional` decorator on `insertDatasets` runs all of this inside a transaction, so it's all definitely atomic.  The failure mode if we had tried to recursively associate later would have just been a lot of unnecessary I/O, as the database replaced a bunch of rows with identical rows.